README.md
mech_util
=========

[![DOI](https://zenodo.org/badge/5316746.svg)](https://zenodo.org/badge/latestdoi/5316746)

This utility manipulates a Chemkin-format chemical kinetic mechanism in one of two ways:
- converts reversible reactions to irreversible reactions
- removes PLOG (pressure-log) reactions

`irrev_mech` requires [Python] 3.x and the [SciPy] stack, but no additional libraries (i.e., Chemkin, Cantera). It has only been tested on [Python] 3.5 and 3.6.

Usage
-------

`mech_util` can be used locally via

    $ python -m mech_util [options]

or installed as a package using `pip install --user mech_util` or via `python setup.py install`, and called using

    $ mech_util [options]

Use the option `-h` or `--help` to see the full usage instructions. To generate an irreversible mechanism, from the command line, use `python -m mech_util -c mech_name -t thermname` where `mech_name` and `therm_name` are the names of the chemical kinetics reaction mechanism file and thermodynamic database, e.g.:

    $ python -m mech_util --model mech.dat --thermo therm.dat --remove_irrev

The new model file has the name `mech_irrev.txt`.

The default temperature range used for parameter fitting is 300 K to 5000 K. This can be changed by specifying the `-r` or `--range` command line option, e.g.:

    $ python -m mech_util --model mech.dat --thermo therm.dat --remove_irrev --temp_range 1000 3000

License
-------

`mech_util` is released under the MIT license, see LICENSE for details.

Citation
--------
If you use this software as part of a scholarly publication, please cite the software directly using the DOI in the badge at the top of this file (and found here: https://zenodo.org/badge/latestdoi/5316746).


Further Reading
---------------

`mech_util` converts reversible reactions into irreversible reactions by fitting reverse Arrhenius coefficients using a nonlinear least-squares minimization.
The three reverse Arrhenius coefficients are determined using a nonlinear least-squares minimization, using the [SciPy] function `scipy.optimize.leastsq`. The fit is performed using a large number of calculated reverse rate coefficients, densely sampled over the specified temperature range. The initial guess for this minimization comes from an analytical fit to three temperatures (the high and low values of the range, plus a midpoint). See the appendix of Niemeyer and Sung's paper for more details:

* KE Niemeyer and CJ Sung. "Accelerating moderately stiff chemical kinetics in reactive-flow simulations using GPUs." *J. Comput. Phys.*, 256:854-871, 2014. doi:[10.1016/j.jcp.2013.09.025](http://dx.doi.org/10.1016/j.jcp.2013.09.025)

The use of the least-squares minimization to obtain better fits was suggested by Taylor et al.:

 * BD Taylor, DA Schwer, and A Corrigan. "Implementation of Thermochemistry and Chemical Kinetics in a GPU-based CFD Code." 53rd AIAA Aerospace Sciences Meeting, Kissimmee, FL, January 2015. doi:[10.2514/6.2015-0842](http://dx.doi.org/10.2514/6.2015-0842)

`mech_util` "de-plogs" models by using a given pressure to either choose the Arrhenius
parameters at the nearest pressure, or by interpolating parameter values.

Author
------

Created by [Kyle Niemeyer](https://niemeyer-research-group.github.io). Email address: [kyle.niemeyer@oregonstate.edu](mailto:kyle.niemeyer@oregonstate.edu)


[Python]: http://python.org/
[SciPy]: http://scipy.org/


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/)
and this project adheres to [Semantic Versioning](http://semver.org/).

## [Unreleased]
### Added
- Added module for de-plogging reactions in a model

### Fixed

### Changed
- Changed package name to "mech_util" to reflect broader capabilities
- Moved write_mech function to separate module
- Removed claim of Python 2.x compatibility


## [2.1.0] - 2016-12-24
### Added
- Added CHANGELOG
- Added single functional test comparing overall results against blessed output
- Added option to name output file

### Changed
- Reactions are now processed in parallel using multiprocessing


## [2.0.0] - 2016-03-29
### Added
- Added Python 3 compatibility

### Fixed
- Minor fixes to install guide in README

### Changed
- Restructured as package from simple module
- Changed command-line options


## 1.0.0 - 2015-11-04
### Added
- First actual release using versioning

### Fixed
- Fixed issue printing third-body reactions

### Changed
- From earlier versions, switched from analytical solution for coefficients
based on three temperatures to least-squares fit based on densely sampled
temperatures


 [Unreleased]: https://github.com/kyleniemeyer/irrev_mech/compare/v0.0.0...HEAD
 [2.1.0]: https://github.com/kyleniemeyer/irrev_mech/compare/v2.0.0...v2.1.0
 [2.0.0]: https://github.com/kyleniemeyer/irrev_mech/compare/v1.0.0...v2.0.0


mech_util/mech_util.py
"""Driver function for package
"""

import sys
import os
from argparse import ArgumentParser

from ._version import __version__
from .irrev_mech import convert_mech_irrev
from .remove_plog import remove_plog_reactions


def mech_util(argv):
    """
    """
    parser = ArgumentParser(description='mech_util: Kinetic model utility.')

    parser.add_argument(
        '-m', '--model',
        help='input model filename for conversion (e.g., "mech.dat").',
        type=str,
        required=True
        )
    parser.add_argument(
        '--thermo',
        help='thermodynamic data filename (only necessary for Chemkin files).',
        type=str,
        default=None
        )
    
    parser.add_argument(
        '--remove_irrev',
        help='Flags conversion of all reversible reactions to irreversible only',
        action='store_true',
        default=False,
        )

    parser.add_argument(
        '--remove_plog',
        help='Flags removal of PLOG reactions',
        action='store_true',
        default=False,
        )
    
    parser.add_argument(
        '--method',
        help='Method for removing PLOG reactions (interpolate, or use nearest)',
        choices=['interpolate', 'nearest'],
        type=str,
        default='nearest',
        )

    parser.add_argument(
        '--pressure',
        help='Specifies pressure for removal of PLOG reactions',
        type=str,
        nargs='+',
        default='1.0 atm',
        )

    parser.add_argument(
        '--temp_range',
        help='Specifies lower and upper temperature limits for fitting irreversible reactions',
        nargs=2,
        type=float,
        default=[300., 5000.],
        )

    parser.add_argument(
        '--num_threads',
        help=('Number of CPU cores to use for working in parallel. '
              'If no number, then use available number of cores minus 1.'
              ),
        nargs='?',
        const=0,
        default=1,
        type=int
        )
    
    parser.add_argument(
        '--output',
        help='output file for new model',
        type=str,
        default='mech.out'
        )

    parser.add_argument(
        '-V', '--version',
        action='store_true',
        help='Show the version of mech_util and quit'
        )

    if len(argv) == 0:
        parser.print_help()
        sys.exit(1)

    args = parser.parse_args()

    if args.version:
        print('mech_util {version} from {path} ()'.format(
            version=__version__,
            path=os.path.abspath(os.path.dirname(__file__))
            ))
        sys.exit(0)
    
    if not args.remove_irrev and not args.remove_plog:
        parser.error('Need to specify either remove_irrev or remove_plog')

    if args.remove_irrev:
        convert_mech_irrev(
            args.model, args.thermo, args.temp_range, 
            args.output, args.num_threads
            )
    
    if args.remove_plog:
        pressure = ' '.join(args.pressure)
        remove_plog_reactions(
            args.model, args.thermo, pressure=pressure, 
            output_file=args.output, method=args.method
            )
